File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def fun(self, nums, ind, dp):
        if ind >= len(nums): return 1000000
        if ind == len(nums)-1: return 0
        if dp[ind] != -1: return dp[ind]
        ans = 1000000
        for i in range(1, nums[ind]+1):
            t = 1+self.fun(nums, ind+i, dp)
            ans = min(ans, t)

        dp[ind] = ans
        return ans

    def jump(self, nums: List[int]) -> int:
        return self.fun(nums, 0, [-1]*len(nums))

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    int fun(int a[], int ind, Integer dp[])
    {
        if(ind>=a.length)
            return 100000000;
        if(ind == a.length-1)
            return 0;
        if(dp[ind]!=null)
            return dp[ind];
        int ans = 100000000;
        for(int i=1; i<=a[ind]; i++)
        {
            int temp = 1 + fun(a, ind+i, dp);
            ans = Math.min(ans, temp);
        }
        return dp[ind] = ans;
    }
    public int jump(int[] nums) {
        Integer dp[] = new Integer[nums.length];
        return fun(nums, 0, dp);
    }
}